modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/model/ReferralTrafficChannel.java

```diff
@@ -0,0 +1,191 @@
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.json.JSONUtil;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.ResourceBundleUtil;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.ResourceBundle;
+import java.util.stream.Stream;
+
+/**
+ * @author David Arques
+ */
+public class ReferralTrafficChannel implements TrafficChannel {
+
+	public static ReferralTrafficChannel withError() {
+		return new ReferralTrafficChannel(
+			true, Collections.emptyList(), Collections.emptyList(), 0, 0);
+	}
+
+	public ReferralTrafficChannel(
+		List<ReferringURL> referringDomains, List<ReferringURL> referringPages,
+		long trafficAmount, double trafficShare) {
+
+		this(
+			false, referringDomains, referringPages, trafficAmount,
+			trafficShare);
+	}
+
+	@Override
+	public boolean equals(Object object) {
+		if (this == object) {
+			return true;
+		}
+
+		if (!(object instanceof ReferralTrafficChannel)) {
+			return false;
+		}
+
+		ReferralTrafficChannel directTrafficChannel =
+			(ReferralTrafficChannel)object;
+
+		if (Objects.equals(
+				getHelpMessageKey(),
+				directTrafficChannel.getHelpMessageKey()) &&
+			Objects.equals(
+				_trafficAmount, directTrafficChannel._trafficAmount) &&
+			Objects.equals(_trafficShare, directTrafficChannel._trafficShare)) {
+
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public String getHelpMessageKey() {
+		return "this-is-the-number-of-page-views-generated-by-people-coming-" +
+			"to-your-page-from-other-sites-which-are-not-search-engine-pages-" +
+				"or-social-sites";
+	}
+
+	@Override
+	public String getName() {
+		return "referral";
+	}
+
+	public List<ReferringURL> getReferringDomains() {
+		return _referringDomains;
+	}
+
+	public List<ReferringURL> getReferringPages() {
+		return _referringPages;
+	}
+
+	@Override
+	public long getTrafficAmount() {
+		return _trafficAmount;
+	}
+
+	@Override
+	public double getTrafficShare() {
+		return _trafficShare;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(
+			getHelpMessageKey(), getName(), _trafficAmount, _trafficShare);
+	}
+
+	@Override
+	public JSONObject toJSONObject(
+		Locale locale, ResourceBundle resourceBundle) {
+
+		JSONObject jsonObject = TrafficChannelUtil.toJSONObject(
+			_error,
+			ResourceBundleUtil.getString(resourceBundle, getHelpMessageKey()),
+			getName(), ResourceBundleUtil.getString(resourceBundle, getName()),
+			_trafficAmount, _trafficShare);
+
+		if (ListUtil.isEmpty(_referringDomains) &&
+			ListUtil.isEmpty(_referringPages)) {
+
+			return jsonObject;
+		}
+
+		return jsonObject.put(
+			"details",
+			JSONUtil.put(
+				"referringDomains", _getReferringDomainsJSONArray()
+			).put(
+				"referringPages", _getReferringPagesJSONArray()
+			));
+	}
+
+	@Override
+	public String toString() {
+		JSONObject jsonObject = TrafficChannelUtil.toJSONObject(
```

Comment:
Better if we call the inner toJSONObject method

Reply:
I'd like to do that, but I'd need a `ResourceBundle` for that or passing a null and handling it within  toJSONObject() which I don't feel comfortable, because the toJSONObject really needs a non-null `ResourceBundle`. Maybe I'm missing other options.